src/shared/port-provider.js

```diff
@@ -0,0 +1,269 @@
+    if (channel === 'host-sidebar' && port === 'host') {
+      return this._hostSidebarChannel.port1;
+    }
+
+    return null;
+  }
+
+  /**
+   * Initiate the listener of port requests by other frames.
+   */
+  _listen() {
+    /** @type {Array<{allowedOrigin: string, channel: Channel, port: Port}>} */
+    ([
+      {
+        allowedOrigin: '*',
+        channel: 'guest-host',
+        port: 'guest',
+      },
+      {
+        allowedOrigin: '*',
+        channel: 'guest-sidebar',
+        port: 'guest',
+      },
+      {
+        allowedOrigin: this._hypothesisAppsOrigin,
+        channel: 'host-sidebar',
+        port: 'sidebar',
+      },
+      {
+        allowedOrigin: this._hypothesisAppsOrigin,
+        channel: 'notebook-sidebar',
+        port: 'notebook',
+      },
+    ]).forEach(({ allowedOrigin, channel, port }) => {
+      this._listeners.add(window, 'message', event =>
+        this._handlePortRequest(/** @type {MessageEvent} */ (event), {
+          allowedMessage: {
+            channel,
+            port,
+            source,
+            type: 'request',
+          },
+          allowedOrigin,
+        })
+      );
+    });
+  }
+
+  /**
+   * @typedef Options
+   * @prop {Message} allowedMessage - the request `MessageEvent` must match this
+   *   object to grant the port.
+   * @prop {string} allowedOrigin - the origin in the `MessageEvent` must match
+   *   this value to grant the port. If '*' allow all origins.
+   */
+
+  /**
+   * Checks the `postMessage` origin and message.
+   *
+   * @param {MessageEvent} event
+   * @param {Options} options
+   */
+  _isValidRequest({ data, origin, source }, { allowedMessage, allowedOrigin }) {
+    if (allowedOrigin !== '*' && origin !== allowedOrigin) {
+      return false;
+    }
+
+    if (
+      // `source` can be of type Window | MessagePort | ServiceWorker.
+      // The simple check `source instanceof Window`` doesn't work here.
+      // Alternatively, `source` could be casted `/** @type{Window} */ (source)`
+      !source ||
+      source instanceof MessagePort ||
+      source instanceof ServiceWorker
+    ) {
+      return false;
+    }
+
+    if (!isMessageEqual(data, allowedMessage)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Send (1) the requested port via `frame.postMessage` (the origin is set
+   * to match the allowedOrigin) and (2) the reciprocal port, if one is provided,
+   * to the `sidebar` frame using `host-sidebar(channel).host(port).postMessage`
+   *
+   * @param {MessageEvent} event
+   * @param {Options & {port: MessagePort, reciprocalPort? : MessagePort}} options
+   */
+  _sendPort(event, { allowedMessage, port, reciprocalPort }) {
+    const message = { ...allowedMessage, type: 'offer' };
+
+    const source = /** @type {Window} */ (event.source);
+
+    source.postMessage(message, event.origin, [port]);
+
+    if (reciprocalPort) {
+      if (['notebook-sidebar', 'guest-sidebar'].includes(message.channel)) {
```

Comment:
Currently, to open the notebook, the sidebar sends a RPC message to the host. The host uses the emitter to communicate to the `Notebook` class, which causes the iframe containing the notebook app to display.

The notebook is currently opened via the sidebar (could change in the the future), so it makes sense to me that the notebook initiates the request. 

Adding a direct notebook-sidebar channel will enable to change groups (in either apps) without reloading their iframes, which is a slow process.

Reply:
When the notebook is opened, my understanding is that two things would happen:

1. Sidebar sends request to host to open notebook
2. Sidebar sends request to notebook to sync groups or other view state

Step 1 will use the sidebar <-> host channel, but step 2 needs the sidebar <-> notebook channel. My thinking when I read this initially was that because the sidebar initiates step 2, it would be the one requesting creation of the sidebar <-> notebook channel. When the notebook opens, it would then request its end of the channel and process any messages that happen to already have been sent on the channel.

Does this match your thinking or were you thinking of a different flow?